core/src/test/scala/integration/kafka/api/BaseProducerSendTest.scala

```diff
@@ -302,6 +305,98 @@ abstract class BaseProducerSendTest extends KafkaServerTestHarness {
   }
 
   /**
+   * testSendBeforeAndAfterPartitionExpansion checks the partitioning behavior before and after partitions are added
+   *
+   * The specified partition-ids should be respected
+   */
+  @Test
+  def testSendBeforeAndAfterPartitionExpansion() {
+    val producer = createProducer(brokerList)
+
+    try {
+      // create topic
+      val leaders = TestUtils.createTopic(zkUtils, topic, 1, 2, servers)
+      val partition0 = 0
+
+      // make sure leaders exist
+      var leader0 = leaders(partition0)
+      assertTrue("Leader for topic \"topic\" partition 0 should exist", leader0.isDefined)
+
+      var responses0 =
+        for (i <- 1 to numRecords)
+        yield producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, partition0, null, ("value" + i).getBytes))
+      var futures0 = responses0.toList
+      futures0.foreach(_.get)
+      for (future <- futures0)
+        assertTrue("Request should have completed", future.isDone)
+
+      // make sure all of them end up in the same partition with increasing offset values
+      for ((future, offset) <- futures0 zip (0 until numRecords)) {
+        assertEquals(offset.toLong, future.get.offset)
+        assertEquals(topic, future.get.topic)
+        assertEquals(partition0, future.get.partition)
+      }
+
+      // Trying to send a record to a partition beyond topic's partition range before adding the partition should fail.
+      val partition1 = 1
+      try {
+        producer.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, partition1, null, "value".getBytes))
+        fail("Should not allow sending a record without topic")
+      } catch {
+        case ke: KafkaException => // this is ok
+        case e: Throwable => fail("Only expecting KafkaException", e)
+      }
+
+      AdminUtils.addPartitions(zkUtils, topic, 2)
+      // wait until leader is elected
+      leader0 = waitUntilLeaderIsElectedOrChanged(zkUtils, topic, 0)
+      val leader1 = waitUntilLeaderIsElectedOrChanged(zkUtils, topic, 1)
+      val leader0FromZk = zkUtils.getLeaderForPartition(topic, 0).get
+      val leader1FromZk = zkUtils.getLeaderForPartition(topic, 1).get
+      assertEquals(leader0.get, leader0FromZk)
+      assertEquals(leader1.get, leader1FromZk)
```

Comment:
Do we need any of the lines after `addPartitions`? Seems like the `waitUntilMetadataIsPropagated` is all that we need for this test.